keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,176 @@
+kind: ConfigMap
+metadata:
+  name: example
+  options:
+    fieldManager: jenkins
+data:
+  k: v
+```
+
+### User Stories [optional]
+
+#### Git Managed Configuration
+
+Alice has a repo containing her current set of Kubernetes manifests as yaml in git.
+The repo gets managed using Gitflow.
+Master branch is automatically applied to the production Kubernetes cluster by a push triggered Jenkins job.
+Staging branch is automatically applied to the staging Kubernetes cluster by a push triggered Jenkins job.
+All members of her team file pull requests against the repo to bring changes into staging and production.
+
+Aside from that, Alice and her team got services that directly manage their objects, like controllers and webhooks,
+as well as services that updated labels for some objects.
+
+They want to make use of the new Server Side Apply, so concurrent changes to objects from different actors are possible without unwanted overlap/conflicts.
+
+To do so, assuming the changes in this KEP, Alice would add the newly introduced field to all manifests in their config management repo. The field would be set to e.g. `configManagement` (or their repo name).
+Additionally, she would update the manifests sent by different actors (like controllers) to contain the field with a value like `controllerName`. Note that this might already be the case for some through the defaulting to the user-agent inside the apply and update pathway.
+
+If now the team updates their repo with a change to their manifests, Jenkins again applies them as usual. On the apiserver side though, the fieldManager set in the applied manifest gets used and all fields set now get owned by the `configManagement` fieldManager.
+
+Again as usual other actors step in to change the objects. They as well use their fieldManager value either set in the manifest file they apply, or based on their user-agent/configuration.
+One of those actors is setting annotations for every service with external information.
+As the actor now only updates annotations not already owned by the `configManagement` fieldManager, applying them will cause the actor to own them.
+
+All actors now can coexist and manage their set of fields without interfering with each other.
+
+#### Kustomize
+
+In reference to the above workflow, it would be possible to override `options.fieldManager` for differen Kustomize overlays and therefor separate ownership depending on the actor currently active.
+
+This means, updating single fields by using a Kustomize overlay (like updating images, labels or annotations) as part of a CI/CD pipeline could easily contain the fieldManager information without any changes to Kustomize (or similar tools) itself.
+
+### Risks and Mitigations
+
+One risk might be, that introducing this change would mean there is a field that does not get persisted to storage.
+As a result, setting it causes different behavior on the apiserver, but is not reflected when reading the object back from the apiserver.
```

Comment:
Why is this a risk? I guess something along the lines of something reads the current object and stores it, and then... ? But I can't envision a particular path that causes harm?

Reply:
OK here's a specific risk: someone sets this in a pod template (because why not); the controller code creates the pod, but diffs it with the template, and assumes something has gone wrong, and recreates the pod.

(this would be a bug in the controller but we've hit it before and it could happen again)

Reply:
The risk here is user confusion.
At least that was what I aimed at when I wrote the section, not sure where it got lost.
So the user might not expect that the field is not persisted and proper naming and documentation would be the proposed fix.